src/prefect/runner/storage.py

```diff
@@ -295,27 +341,52 @@ async def _clone_repo(self):
         # Add the clone command and its parameters
         cmd += ["clone", repository_url]
 
-        if self._branch:
-            cmd += ["--branch", self._branch]
         if self._include_submodules:
             cmd += ["--recurse-submodules"]
 
         # This will only checkout the top-level directory
         if self._directories:
             cmd += ["--sparse"]
 
-        # Limit git history and set path to clone to
-        cmd += ["--depth", "1", str(self.destination)]
+        if self._commit_sha:
+            cmd += [str(self.destination)]
 
-        try:
-            await run_process(cmd)
-        except subprocess.CalledProcessError as exc:
-            # Hide the command used to avoid leaking the access token
-            exc_chain = None if self._credentials else exc
-            raise RuntimeError(
-                f"Failed to clone repository {self._url!r} with exit code"
-                f" {exc.returncode}."
-            ) from exc_chain
+            try:
+                await run_process(cmd)
+            except subprocess.CalledProcessError as exc:
+                # Hide the command used to avoid leaking the access token
+                exc_chain = None if self._credentials else exc
+                raise RuntimeError(
+                    f"Failed to clone repository {self._url!r} with exit code"
+                    f" {exc.returncode}."
+                ) from exc_chain
+
+            # Checkout the specific commit
+            await run_process(
+                ["git", "checkout", self._commit_sha],
+                cwd=self.destination,
+            )
+            self._logger.debug(f"Successfully checked out commit {self._commit_sha}")
+
+        else:
+            if self._branch:
+                cmd += ["--branch", self._branch]
+
+            # Limit git history and set path to clone to
+            if not self._commit_sha:
+                cmd += ["--depth", "1"]
+
+            cmd += [str(self.destination)]
+
+            try:
+                await run_process(cmd)
+            except subprocess.CalledProcessError as exc:
+                # Hide the command used to avoid leaking the access token
+                exc_chain = None if self._credentials else exc
+                raise RuntimeError(
+                    f"Failed to clone repository {self._url!r} with exit code"
+                    f" {exc.returncode}."
+                ) from exc_chain
```

Comment:
I think this branch can be consolidated with the branch above to dedup some of the code.